include/net/cloud.h

```diff
@@ -0,0 +1,246 @@
+{
+	if (backend == NULL) {
+		return -EINVAL;
+	}
+
+	if (handler == NULL) {
+		return -EINVAL;
+	}
+
+	return backend->api->init(backend, handler);
+}
+
+/**@brief API to uninitialize a cloud backend. Gracefully disconnects
+ *        remote endpoint and releases memory.
+ *
+ * param[in] cloud Pointer to cloud backend structure.
+ */
+static inline int cloud_uninit(const struct cloud_backend *const backend)
+{
+	__ASSERT_NO_MSG(backend != NULL);
+
+	return backend->api->uninit(backend);
+}
+/**@brief Request connection to cloud backend.
+ *
+ * @details The backend is required to expose the socket in use when this
+ *	    function returns. The socket should then be available through
+ *	    backend->config->socket and the application may start listening
+ *	    for events on it.
+ *
+ * param[in] cloud Pointer to a cloud backend structure.
+ *
+ * @return 0 or a negative error code indicating reason of failure.
+ */
+static inline int cloud_connect(const struct cloud_backend *const backend)
+{
+	__ASSERT_NO_MSG(backend != NULL);
+
+	return backend->api->connect(backend);
+}
+
+/**@brief API to disconnect from cloud backend.
+ *
+ * param[in] cloud Pointer to a cloud backend structure.
+ *
+ * @return 0 or a negative error code indicating reason of failure.
+ */
+static inline int cloud_disconnect(const struct cloud_backend *const backend)
+{
+	__ASSERT_NO_MSG(backend != NULL);
+
+	return backend->api->disconnect(backend);
+}
+
+/**@brief API to send data to cloud.
+ *
+ * param[in] cloud Pointer to a cloud backend structure.
+ * param[in] msg Pointer to cloud message structure.
+ *
+ * @return 0 or a negative error code indicating reason of failure.
+ */
+static inline int cloud_send(const struct cloud_backend *const backend,
+		    struct cloud_msg *msg)
+{
+	__ASSERT_NO_MSG(backend != NULL);
+	__ASSERT_NO_MSG(msg != NULL);
+
+	return backend->api->send(backend, msg);
+}
+
+/**
+ * @brief Optional API to ping the cloud's remote endpoint periodically.
+ *
+ * @param[in] backend Pointer to cloud backend.
+ *
+ * @return 0 or a negative error code indicating reason of failure.
+ */
+static inline int cloud_ping(const struct cloud_backend *const backend)
+{
+	__ASSERT_NO_MSG(backend != NULL);
+
+	/* Ping will only be sent if the backend requires it. */
+	if (backend->api->ping != NULL) {
+		return backend->api->ping(backend);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Process incoming data to backend.
+ *
+ * @note This is a non-blocking call.
+ *
+ * @param[in] backend Pointer to cloud backend.
+ *
+ * @return 0 or a negative error code indicating reason of failure.
+ */
+static inline int cloud_input(const struct cloud_backend *const backend)
+{
+	__ASSERT_NO_MSG(backend != NULL);
+
+	return backend->api->input(backend);
+}
+
+struct cloud_backend *cloud_get_binding(const char *name);
+
+#define CLOUD_BACKEND_DEFINE(_name, _api)			               \
+									       \
+	static struct cloud_backend_config UTIL_CAT(_name, _config) =	       \
```

Comment:
I think we need to think thoroughly about whether this is the right direction. Technically it seems fine, but it will limit the cloud engine to a single instance. And I wonder whether it would make sense to maintain two parallel connections to the same cloud from a single device with let's say two different identities. But it might be an overcomplicated use case.

Reply:
Yes, I'm not sure about this either. It's a hybrid of how logger and device API handle their backends and devices. Big downside is that it requires changes to linker script in the -zephyr repo, unless there's a way that I haven't found to do it another way.

The thinking behind it is that it's convenient to use some sort of `cloud_get_binding("CLOUD NAME")` call to get pointer to the backend struct, and allow mulitple backends concurrently. But as you say, only one per cloud solution. There's probably a better way to solve this.

Reply:
Could this https://github.com/zephyrproject-rtos/zephyr/pull/15112 be used to add the relevant sections to the linker script? @jtguggedal

Reply:
I forgot to comment here on it, but I accidently came across this solution and updated the Cloud API to use it. I placed the linker snippet within the subsys itself like this:  https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/795/files#diff-382398a27417821bcb12e888d758bda0R7 
Do you think this is an ok solution, @sigvartmh @rlubos ?